code/datums/voice_announcements.dm

```diff
@@ -0,0 +1,174 @@
+GLOBAL_LIST_EMPTY(voice_announce_list)
+
+/datum/voice_announce
+	var/id
+	var/client/client
+	var/is_ai = FALSE
+	var/open = FALSE
+	var/started_playing = FALSE
+	var/duration = 300
+	var/canceled = FALSE
+	var/was_queried = FALSE
+
+/datum/voice_announce/New(client/client)
+	. = ..()
+	src.client = client
+	id = "[client.ckey]_[sha1("[world.time]-[world.realtime]-[world.timeofday]-[rand()]")]"
+
+/datum/voice_announce/Destroy()
+	GLOB.voice_announce_list -= id
+	. = ..()
+
+/datum/voice_announce/proc/open()
+	if(SScommunications.last_voice_announce_open + 30 > world.time)
+		// Keep cheeky fucks from trying to waste resources by spamming the button
+		return
+	var/url_base = CONFIG_GET(string/voice_announce_url_base)
+	var/dir = CONFIG_GET(string/voice_announce_dir)
+	if(!url_base || !dir)
+		return
+
+	if(is_banned_from(client.ckey, "Voice Announcements"))
+		to_chat(client, "<span class='warning>You are banned from making voice announcements.</span>")
+		return
+
+	SScommunications.last_voice_announce_open = world.time
+
+	GLOB.voice_announce_list[id] = src
+	open = TRUE
+	usr << link(url_base + "[CONFIG_GET(string/serversqlname)]/[id]")
+	addtimer(CALLBACK(src, .proc/timeout1), 15 SECONDS)
+	addtimer(CALLBACK(src, .proc/timeout2), 5 MINUTES)
+
+/datum/voice_announce/proc/timeout1()
+	if(!was_queried && !started_playing)
+		qdel(src)
+
+/datum/voice_announce/proc/timeout2()
+	if(!started_playing)
+		qdel(src)
+
+/datum/voice_announce/Topic(href, href_list)
+	. = ..()
+	if(href_list["stop_announce"] && started_playing && !canceled && check_rights(NONE))
+		SEND_SOUND(world, sound(null, channel = CHANNEL_VOICE_ANNOUNCE))
+		canceled = TRUE
+		log_admin("[key_name(usr)] has canceled the voice announcement")
+		message_admins("[key_name_admin(usr)] has canceled the voice announcement")
+
+
+/datum/voice_announce/proc/do_announce_sound(sound/sound1, sound/sound2, sounds_delay, z_level)
+	started_playing = TRUE
+	// Play for admins
+	for(var/mob/M in GLOB.player_list)
+		if(M.client && M.client.holder && M.can_hear() && (M.client.prefs.toggles & SOUND_ANNOUNCEMENTS))
+			var/turf/T = get_turf(M)
+			if(T.z == z_level)
+				SEND_SOUND(M, sound1)
+	sleep(sounds_delay)
+	for(var/mob/M in GLOB.player_list)
+		if(M.client && M.client.holder && M.can_hear() && (M.client.prefs.toggles & SOUND_ANNOUNCEMENTS))
+			var/turf/T = get_turf(M)
+			if(T.z == z_level)
+				SEND_SOUND(M, sound2)
+	sleep(30)
```

Comment:
Wouldn't it make more sense to do this?
```suggestion
	sleep(duration + 5 SECONDS)
```

Reply:
That's an unreasonably long wait time IMO

Reply:
I'm not too sure about having it played for everyone before admins get to hear the full message tho

Reply:
admins get to hear bad words before everyone else, hit cancel button, sure they'll get a message cut off halfway but whatever right?